pkg/log/dedup/dedup_middleware.go

```diff
@@ -0,0 +1,384 @@
+	"time"
+)
+
+// EmittedAttrKey marks a record that is being emitted by the deduper itself
+// (e.g., during cleanup) so the middleware can skip deduplication to prevent
+// feedback loops.
+const EmittedAttrKey = "__dedup_emitted"
+
+// Default configuration values.
+const (
+	DefaultCacheExpiry        = 5 * time.Minute
+	DefaultMaxCacheSize       = 2000
+	DefaultCleanupInterval    = 1 * time.Minute
+	DefaultDuplicateLogWindow = 1 * time.Minute
+)
+
+const duplicateSummaryMsg = "duplicate_summary"
+
+// excludedHashFields are the attribute keys that should not affect the content hash.
+var excludedHashFields = map[string]bool{
+	"ts":              true, // go-kit timestamp
+	"time":            true, // slog timestamp
+	"caller":          true, // go-kit caller info
+	"source":          true, // slog source info
+	"original.time":   true,
+	"original.source": true,
+	EmittedAttrKey:    true, // internal marker
+}
+
+// Config controls the behavior of the dedup middleware.
+type Config struct {
+	CacheExpiry        time.Duration
+	MaxCacheSize       int
+	CleanupInterval    time.Duration
+	DuplicateLogWindow time.Duration
+}
+
+// Option configures the deduper.
+type Option func(*Config)
+
+// WithCacheExpiry overrides the default cache expiry.
+func WithCacheExpiry(d time.Duration) Option { return func(c *Config) { c.CacheExpiry = d } }
+
+// WithMaxCacheSize overrides the default maximum cache size.
+func WithMaxCacheSize(n int) Option { return func(c *Config) { c.MaxCacheSize = n } }
+
+// WithCleanupInterval overrides the cleanup interval.
+func WithCleanupInterval(d time.Duration) Option { return func(c *Config) { c.CleanupInterval = d } }
+
+// WithDuplicateLogWindow overrides the window to re-log duplicates.
+func WithDuplicateLogWindow(d time.Duration) Option {
+	return func(c *Config) { c.DuplicateLogWindow = d }
+}
+
+// logEntry tracks information about seen log messages for deduplication.
+type logEntry struct {
+	firstSeen  time.Time
+	lastSeen   time.Time
+	count      int
+	lastLogged time.Time
+
+	// For emission on cleanup
+	level   slog.Level
+	message string
+	attrs   []slog.Attr
+	pc      uintptr
+}
+
+// Engine is a stateful deduplication engine. It is safe for concurrent use.
+type Engine struct {
+	// configuration
+	cfg Config
+
+	// runtime state
+	mu          sync.RWMutex
+	cache       map[string]*logEntry
+	lastCleanup time.Time
+	// ensure only one cleanup runs at a time
+	cleanupRunning int32
+
+	// background cleanup machinery
+	ctx    context.Context //nolint:containedctx // Used for background goroutine lifecycle
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	// logger used to emit records on cleanup
+	logger *slog.Logger
+}
+
+// NewEngine creates a new deduplication engine. The provided logger is used to
+// emit summary records on expiration; pass the pipeline's logger so records go
+// through the same handler chain.
+func NewEngine(logger *slog.Logger, opts ...Option) *Engine {
+	cfg := Config{
```

Comment:
Why have a separate struct for `Config` (vs using these as unexpected attributes on engine)

Reply:
You have the same levels if they're struct variables on the engine itself. A common construction would be to make the engine, then pass the engine to the functions.